Destructor of record list created

sagar-sam opened the pull request:
Issue #4956 
Destructor of record list is created which free the memory resources.

mppf commented:
hi @sagar-sam - thanks for giving this a try. You'll need to make sure things are working by running tests - see https://github.com/chapel-lang/chapel/blob/master/doc/developer/bestPractices/TestSystem.rst

A good start would be to verify no failures when you run

    ./util/start_test test/release/examples/

mppf commented:
Also please have a look at https://github.com/chapel-lang/chapel/blob/master/doc/developer/bestPractices/ContributorInfo.rst

sagar-sam commented:
Thanks for your suggestion @mppf 
I will look into it.

sagar-sam commented:
I executed    ./util/start_test test/release/examples/
The summary is

[Test Summary - 170309.175442]
[Error matching program output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/learnChapelInYMinutes]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/opaque]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/randomNumbers]
[Error matching program output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/associative]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/benchmarks/hpcc/stream-ep]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/benchmarks/hpcc/stream]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/benchmarks/hpcc/variants/stream-promoted]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/users-guide/base/constParam (compopts: 1)]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/users-guide/base/constParam (compopts: 2)]
[Warning: Could not find chpldoc, skipping test /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/chpldoc.doc]
[Summary: #Successes = 139 | #Failures = 9 | #Futures = 0 | #Warnings = 1 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
[END]

sagar-sam commented:
The same failures occurred when i executed the above file on branch master

mppf commented:
Right, so you have to look into why those tests are failing. Try compiling one from the command line to see the error.

sagar-sam commented:
I created branch 4956 and modify it and got the same failures as branch master which have the unmodified code.

mppf commented:
you're saying you see these errors on master?

mppf commented:
Sorry, for some reason I missed it when you already said that. Could you show the error compiling one of them? Does `make clobber; make` resolve the issue?

sagar-sam commented:
Yes.

bradcray commented:
Just wanted to note that leaked `listNode`s are currently in our top-10 category of memory leaks in nightly testing, so it'd be great to get this in for the release to see if it helps with that category.

````
====================
Leaked Memory Report
==============================================================
Number of leaked allocations
              Total leaked memory (bytes)
                           Description of allocation
==============================================================
       14973      27419030  array elements
         186       4253766  io buffer or bytes
       14151       1471752  domain(1,int(64),false)
        4963        714672  domain(2,int(64),false)
        3434        714272  [domain(1,int(64),false)] chpl_TableEntry(int(64))
        4006        576864  domain(2,int(64),true)
        4050        550800  BlockCyclic1dom(int(64),int(64),false)
        3434        521968  DefaultAssociativeDom(int(64),true)
       18723        449352  listNode(BaseArr)
        2030        422240  [domain(1,int(64),false)] chpl_TableEntry(2*int(64))
````

Tagging @ronawho as we were wondering about this on the way home last night (specifically, Elliot, this represents the largest quantity of leaked items.  Note that that's not the same as you asked -- what leak is impacting the largest number of tests... that's a harder question to answer).

sagar-sam commented:
These type of errors are there in branch master and same in branch 4956


< A populated associative domain of strings: {Alice, Bob, Clark, Dana, Ellen, Frank}
---
> A populated associative domain of strings: {Dana, Alice, Frank, Clark, Bob, Ellen}
7c13
< Removed 'Bob' and 'Frank', added 'Robert': {Alice, Clark, Dana, Ellen, Robert}
---
> Removed 'Bob' and 'Frank', added 'Robert': {Dana, Alice, Clark, Robert, Ellen}
18,19c24,25
< Some primes in the Fibonacci sequence: {13, 2, 3, 5}
< Some primes not in the Fibonacci sequence: {11, 17, 7}
---
> Some primes in the Fibonacci sequence: {5, 13, 3, 2}
> Some primes not in the Fibonacci sequence: {7, 17, 11}
21,22c27,28
< Women = {Alice, Dana, Ellen}
< Men = {Clark, Gary, John, Mark, Robert}
---
> Women = {Dana, Alice, Ellen}
> Men = {John, Gary, Clark, Robert, Mark}
[Error matching program output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/associative]

bradcray commented:
These look to me like nondeterministic orderings based on hashing differences (though I'm not sure why you'd be seeing them and we don't).  Best guess:  Does the test/release/examples/primers/associative.prediff in your copy of the repo not have executable permissions?

sagar-sam commented:
How do I check this??

sagar-sam commented:
These time i got 6 failures

[Test Summary - 170309.235800]
[Error matching program output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/learnChapelInYMinutes]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/opaque]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/randomNumbers]
[Error matching program output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/associative]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/users-guide/base/constParam (compopts: 1)]
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/users-guide/base/constParam (compopts: 2)]
[Warning: Could not find chpldoc, skipping test /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/chpldoc.doc]
[Summary: #Successes = 142 | #Failures = 6 | #Futures = 0 | #Warnings = 1 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
[END]

mppf commented:
http://stackoverflow.com/questions/22122643/how-to-check-if-a-regular-file-is-executable-or-not-in-linux

bradcray commented:
Most of these tests you list also rely on prediff's to filter out nondeterministic behavior.  The few that stick out for me as being surprising are the two constParam cases and the randomNumbers cases.  In what way are those ones failing?

sagar-sam commented:
[Executing diff randomNumbers.good randomNumbers.comp.out.tmp]
1,15c1,3
< randsSeeded = 0.805971 0.719587 0.725396 0.645748 0.903321 0.383077 0.144684 0.991891 0.124353 0.259023
<
< rand16s = 52820 47158 47539 42319 59200 25105 9481 65004 8149 16975
<
< true
< true
< true
< true
< true
< 0.903321
< 0.383077
< 0.144684
< 0.991891
< 0.124353
< 0.259023
---
> randomNumbers.chpl:53: error: unresolved call 'RandomStream.init(type real(64), int(64))'
> $CHPL_HOME/modules/standard/Random.chpl:574: note: candidates are: RandomStream(seed: int(64) = SeedGenerator.currentTime, param parSafe: bool = true, type eltType = real(64))
> $CHPL_HOME/modules/standard/Random.chpl:535: note:                 RandomStream.init(type eltType, param parSafe: bool, seed, PCGRandomStreamPrivate_lock$, PCGRandomStreamPrivate_rngs, PCGRandomStreamPrivate_count)
[Error matching compiler output for /home/sagar/Desktop/sagar/chapel/test/release/examples/primers/randomNumbers]

bradcray commented:
(You'll find that your terminal output cut-and-pastes will format better on GitHub if you surround them with lines made up of four back-tics  (`) :

````
Preserve original formatting
````

bradcray commented:
@mppf may have a better sense, but I the diff above suggests to me that your copy of test/release/examples is not in synch with your copy of the modules/ directory.  Are you running both from a single repository, based on master?  Is there any chance you're running an older version of the Chapel compiler (e.g., the official 1.14.0 release?)

sagar-sam commented:
yes, i have the 1.14.0

mppf commented:
you need to be using master. See https://github.com/chapel-lang/chapel/blob/master/doc/developer/bestPractices/GitCheatsheet.rst

sagar-sam commented:
Can you please tell me the recent version of chapel compiler and provide me the link for the same
Thank you

mppf commented:
you have to clone the master git repository, e.g.
```
git clone https://github.com/chapel-lang/chapel.git
```

but there is more to do. Please see the document I linked to.

sagar-sam commented:
Thanks everyone for tolerating me,
finally there are no failures.
Also first I updated my master branch then merge it into my 4956
Thanks once again.

bradcray commented on modules/standard/List.chpl:
```diff
@@ -186,6 +186,13 @@ record list {
+  /*
+    Destructor
+   */
+  proc ~list(){
```
We've recently started replacing destructors like this with deinit(), so if you could change this over to:

````chapel
proc deinit() {
````

bradcray replied:
Also, there's a TODO above on destroy() which looks like it could be removed now.

bradcray commented:
It would be good to add a new test to the Chapel testing system (say under test/modules/standard/list) that creates a list and destroys it (say within a block statement or a function), measuring the memory delta across that code block using the Memory module:  http://chapel.cray.com/docs/latest/modules/standard/Memory.html?highlight=memory
as a means of making sure that the List class cleans up after itself as expected.

sagar-sam commented:
Ok, I will look into it

sagar-sam commented:
I added memory.chpl and memory.good in directory test/modules/standard/list/
anything else to do?

bradcray commented:
I'm taking a look at this and am starting to run some testing on it.

bradcray commented on test/modules/standard/List/memory.good:
```diff
@@ -0,0 +1,3 @@
+0
+704
+584
```
Hi @sagar-sam : I was wondering why the final memory was not '0' since the List's memory ought to be completely reclaimed and, using printMemAllocsByType(), found that it's due to the I/O that's done within the routine (which is ultimately cleaned up before the program exits, but unfortunately not before your final memoryUsed() call).

For that reason, would you please change this to move all I/O after all of the memoryUsed() calls are made such that the .good file can show that 0 bytes are used after output() completes?  E.g., capture the initial memoryUsed() in a variable.

BTW, while the memoryUsed() call within output() is interesting, I wouldn't include it in the test because the size of the listNode class could vary from one compiler/platform to another, which would make this test non-portable.  This will also trivially remove another I/O call.

Essentially, the goal of the above changes is to have the .good file read:

````
0
0
````

Let me know if this doesn't make sense.

Meanwhile, I'm going to fire off some testing on this branch to make sure that nothing breaks and will also see what impact it has on memory leaks on a subset of the tests.

sagar-sam commented:
Let me try

sagar-sam commented:
The memory is used in the calling of the function.
so, i stored the output of memoryUsed in a variable inside the function call and on subtracting final memoryused from that var results 0.

bradcray commented on test/modules/standard/List/memory.chpl:
```diff
@@ -0,0 +1,19 @@
+var mem : uint;
+proc output()
+{
+	mem=memoryUsed();
```
Oh, I forgot to mention earlier.  This test needs to be run with --memTrack, right?  If so, you should add a memory.execopts file specifying that:

````
--memTrack
````

Otherwise, I think running util/start_test on this test will fail?

sagar-sam replied:
Yes, this test run with --memTrack.

bradcray commented on test/modules/standard/List/memory.chpl:
```diff
@@ -0,0 +1,19 @@
+memTrack;
+writeln(memoryUsed());
+output();
+writeln(mem-memoryUsed());
```
This subtraction is backwards.  If memory were leaked (as it is on master), memoryUsed() would be > mem.

bradcray commented:
This passed full testing, so I think that once the .execopts file is added, this is ready to merge.  Also, I just noted one minor bug in the new test that should be addressed.

Unfortunately, this change did not close any of the listNode() memory leaks in test/release/examples (note @ronawho), so that requires more investigation.  But this is still an improvement over master (e.g., the new test fails on master prior to this change).

sagar-sam commented:
Is it ok now?

bradcray commented:
Looks good, thanks @sagar-sam!

sagar-sam commented:
Most Welcome and Thanks for the merging  @bradcray

bradcray commented:
This unfortunately turned out to have some memory corruption problems in testing last night that my testing hadn't caught, but which had adverse effects in different testing situations (and which valgrind demonstrated once I used that).  For example, see:

https://sourceforge.net/p/chapel/mailman/message/35716321/

The cause of the problem was that when we copied 'list' records, we were doing a shallow copy, and so sharing the listNode objects between the two records.  But then each record would try to free the list, causing problems.

I have a proposed fix for this in PR #5632 which adds an assignment overload and chpl__initCopy() overload (a placeholder for copy constructors for generic types) to do a deep copy for lists, which seems like the appropriate default semantics (though that's up for potential discussion).

There was also one simple problem in which the new test memory.chpl resulted in an ambiguity with the standard Memory.chpl module on systems that are case-insensitive with respect to filenames.  I renamed the test to address this in PR #5630.

Hopefully this will clear up these test configurations tomorrow.

bradcray commented:
Hi @sagar-sam -- I'm updating the CONTRIBUTORS.md file for our upcoming release and wanted to make sure you got credit for this work.  Would you be able to provide me with your name?  You can email me at one of the Chapel mailing lists (e.g., chapel_info@...) if you'd prefer not to put it here.

I'm also realizing belatedly that I should've asked you to fill out an ICLA before accepting your PR.  Would you be able to fill one out for us?  See:  https://github.com/chapel-lang/chapel/tree/master/doc/developer/contributorAgreements/  for details.

Thanks again for your help improving Chapel!

sagar-sam commented:
Hello @bradcray 
My name Sagar Khatri
I filled the contributor agreement, where do I have to submit?

bradcray commented:
Hi @sagar-sam -- Great!  See the README (which  moved this morning, unfortunately, to get the submission address:  https://github.com/chapel-lang/chapel/tree/master/doc/rst/developer/contributorAgreements

sagar-sam commented:
Hi @bradcray 
I submitted the contributor agreement

bradcray commented:
Thanks!
